lookout/style/format/rules.py

```diff
@@ -173,48 +166,43 @@ def _get_composite(feature_extractor: FeatureExtractor, labels: Tuple[int, ...])
         feature_extractor.labels_to_class_sequences.append(labels)
         return len(feature_extractor.labels_to_class_sequences) - 1
 
-    def _get_new_rule(self, feature_extractor: FeatureExtractor, labels: Tuple[int, ...],
-                      new_rules: MutableMapping[Tuple[int, ...], int]) -> int:
-        if labels in new_rules:
-            return new_rules[labels]
-        self._rules.append(Rule(tuple(),
-                                RuleStats(cls=self._get_composite(feature_extractor, labels),
-                                          conf=1., support=1)))
-        rule_i = len(self.rules) - 1
-        new_rules[labels] = rule_i
-        return rule_i
-
-    def _set_new_rule(self, feature_extractor: FeatureExtractor, labels: Tuple[int, ...],
-                      new_rules: MutableMapping[Tuple[int, ...], int], winners: numpy.ndarray,
-                      y: numpy.ndarray, y_i: int) -> None:
-        rule = self._get_new_rule(feature_extractor, tuple(labels), new_rules)
-        winners[y_i] = rule
-        y[y_i] = self._rules[rule].stats.cls
-
-    def harmonize_quotes(self, X: numpy.ndarray, y_pred: numpy.ndarray,
+    def harmonize_quotes(self, y_pred: numpy.ndarray,
                          vnodes_y: Sequence[VirtualNode], vnodes: Sequence[VirtualNode],
                          winners: numpy.ndarray, feature_extractor: FeatureExtractor
-                         ) -> Tuple[numpy.ndarray, numpy.ndarray]:
+                         ) -> Tuple[numpy.ndarray, numpy.ndarray, List[Rule]]:
         """
         Post-process predictions to correct mis-matched quotes.
 
         To do so, we consider only the tuples (', STRING, ') or (", STRING, ") in the input. We
         then create fake rules as needed (because a rule going from the input to the corrected
         quote might not exist in the trained rules).
 
-        :param X: Feature matrix.
         :param y_pred: Predictions to correct.
         :param vnodes_y: Sequence of the predicted virtual nodes.
         :param vnodes: Sequence of virtual nodes representing the input.
         :param winners: Indices of the rules that were used to compute the predictions.
         :param feature_extractor: FeatureExtractor used to extract features.
-        :return: Updated y and winners.
+        :return: Updated y, winners and rules.
         """
         quotes_classes = {CLASS_INDEX[CLS_DOUBLE_QUOTE], CLASS_INDEX[CLS_SINGLE_QUOTE]}
+        processed_rules = list(self.rules)
         processed_y = y_pred.copy()
         processed_winners = winners.copy()
-        y_indices = {id(vnode): i for i, vnode in enumerate(vnodes_y)}
         new_rules = {}
+
+        def append_new_rule(labels: Tuple[int, ...], y_i: int, conf: float, support: int) -> None:
+            if labels in new_rules:
```

Comment:
this does not assign the correct stats if a previous rule on the same labels was created with different stats.

Reply:
Yeah. do you think it is a good idea to create a new rule here too?

Reply:
yes sadly it is necessary to have as many rules as (cls, precision, tuple) comboes :(

Reply:
So there is a case when we can reuse rule. They goes from 
```python
            if labels1[-1] not in quotes_classes or labels3[0] not in quotes_classes:
                append_new_rule(vnode1.y, y_i_1, 1., 1)
                append_new_rule(vnode3.y, y_i_3, 1., 1)
```

so it is better to keep the logic, but chage `new_rules` indexes.  that what I did